File: tree/543-diameter-of-binary-tree.cpp
// Solution 1: recursive
// Compute the diameter of the path which pass
// through the current node while computing
// its height.
//
// N = number of nodes in tree
// O(N) in time (since each node is visited once)
// O(N) in space in worst case (skewed tree)
// O(logN) in space in average
//
// pattern: post order traversal
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int diameter = 0;
        getHeight(root, diameter);
        return diameter;
    }

    int getHeight(TreeNode* node, int & diameter) {
        if (node == NULL)
            return 0;

        int heightLeft = getHeight(node->left, diameter);
        int heightRight = getHeight(node->right, diameter);

        diameter = max(diameter, heightLeft + heightRight
                                + (node->left != NULL)
                                + (node->right != NULL));
        return max(heightLeft + (node->left != NULL),
                   heightRight + (node->right != NULL));
    }
};

// Solution 2: iterative using stack + hash table
// Please refer to Python file, solution 1.
// The idea is same as solution 1, we use hash table
// to store the height at each node.
// Time & Complexity same as solution 1

// Solution 3: optimal recursive solution, improvation
// of solution 1, using the fact that height(null) = -1.
// Think like we also count NULL as a node in tree.
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int diameter = 0;
        getHeight(root, diameter);
        return diameter;
    }

    int getHeight(TreeNode* node, int & diameter) {
        if (node == NULL)
            return -1;

        int heightLeft = getHeight(node->left, diameter);
        int heightRight = getHeight(node->right, diameter);
        diameter = max(diameter, 2 + heightLeft + heightRight);

        return 1 + max(heightLeft, heightRight);
    }
};

// Solution 4: optimal iterative solution, improvation
// of solution 2, using same idea as solution 3
// Please refer to Python file, solution 2

File: practices/leetcode/tree/543-diameter-of-binary-tree.py
# Solution 1: iterative
class Solution:
    def diameterOfBinaryTree(self, root):
        if root is None: return 0
        stack = [root]
        height = {None: 0}
        diameter = 0
        while stack:
            node = stack[-1]
            if (not node.left or node.left in height) and \
               (not node.right or node.right in height):
                # Two children of this node are visited,
                # we can compute the diameter.
                diameter = max(diameter,
                               height[node.left] + height[node.right] \
                             + (node.left != None) + (node.right != None))
                height[node] = max(height[node.left] + (node.left != None),
                                   height[node.right] + (node.right != None))
                stack.pop()
            else:
                if node.left:
                    stack.append(node.left)
                if node.right:
                    stack.append(node.right)
        return diameter

# Solution 2: optimal iterative
# O(N) in space & time
class Solution:
    def diameterOfBinaryTree(self, root):
        if root is None: return 0
        stack = [root]
        height = {None: -1}
        diameter = 0
        while stack:
            node = stack[-1]
            if (not node.left or node.left in height) and \
               (not node.right or node.right in height):
               diameter = max(diameter, 2 + height[node.left] + height[node.right])
               height[node] = max(height[node.left], height[node.right]) + 1
               stack.pop()
           else:
               if node.left: stack.append(node.left)
               if node.right: stack.append(node.right)
        return diameter
